File: pkg/db/package.go
package db

import (
	"database/sql"
	"strconv"
	"strings"

	"github.com/nektro/go-util/util"
	dbstorage "github.com/nektro/go.dbstorage"
	"github.com/nektro/go.etc/dbt"
)

type Package struct {
	ID            int64    `json:"id"`
	UUID          dbt.UUID `json:"uuid" dbsorm:"1"`
	Owner         dbt.UUID `json:"owner" dbsorm:"1"`
	Name          string   `json:"name" dbsorm:"1"`
	CreatedOn     dbt.Time `json:"created_on" dbsorm:"1"`
	Remote        int64    `json:"remote" dbsorm:"1"`
	RemoteID      string   `json:"remote_id" dbsorm:"1"`
	RemoteName    string   `json:"remote_name" dbsorm:"1"`
	Description   string   `json:"description" dbsorm:"1"`
	License       string   `json:"license" dbsorm:"1"`
	LatestVersion string   `json:"latest_version" dbsorm:"1"`
	hookSecret    string   `json:"hook_secret" dbsorm:"1"`
	StarCount     int      `json:"star_count" dbsorm:"1"`
}

// CreatePackage creates a new Package
func CreatePackage(owner *User, name string, remote int64, remoteID, remoteName, description, license string, starcount int) *Package {
	dbstorage.InsertsLock.Lock()
	defer dbstorage.InsertsLock.Unlock()
	//
	id := db.QueryNextID(cTablePackages)
	uid := dbt.NewUUID()
	co := now()
	hooksecret := strings.ToLower(util.RandomString(16))
	n := &Package{id, uid, owner.UUID, name, co, remote, remoteID, remoteName, description, license, "", hooksecret, starcount}
	db.Build().InsI(cTablePackages, n).Exe()
	return n
}

//
//

// Scan implements dbstorage.Scannable
func (v Package) Scan(rows *sql.Rows) dbstorage.Scannable {
	rows.Scan(&v.ID, &v.UUID, &v.Owner, &v.Name, &v.CreatedOn, &v.Remote, &v.RemoteID, &v.RemoteName, &v.Description, &v.License, &v.LatestVersion, &v.hookSecret, &v.StarCount)
	return &v
}

// All queries database for all currently existing Packages
func (v Package) All() []*Package {
	arr := dbstorage.ScanAll(v.b(), Package{})
	res := []*Package{}
	for _, item := range arr {
		res = append(res, item.(*Package))
	}
	return res
}

func (v Package) ByUser(user *User) []*Package {
	arr := dbstorage.ScanAll(v.b().Wh("owner", user.UUID.String()), Package{})
	res := []*Package{}
	for _, item := range arr {
		res = append(res, item.(*Package))
	}
	return res
}

func (v Package) ByUID(ulid dbt.UUID) *Package {
	us, ok := dbstorage.ScanFirst(v.b().Wh("uuid", string(ulid)), Package{}).(*Package)
	if !ok {
		return nil
	}
	return us
}

func (v Package) GetLatest(n int) []*Package {
	arr := dbstorage.ScanAll(v.b().Or("id", "desc").Lm(int64(n)), Package{})
	res := []*Package{}
	for _, item := range arr {
		res = append(res, item.(*Package))
	}
	return res
}

func (v Package) TopStarred(n int) []*Package {
	arr := dbstorage.ScanAll(v.b().Or("star_count", "desc").Lm(int64(n)), Package{})
	res := []*Package{}
	for _, item := range arr {
		res = append(res, item.(*Package))
	}
	return res
}

//
//

func (v *Package) i() string {
	return v.UUID.String()
}

func (v Package) t() string {
	return cTablePackages
}

func (v Package) b() dbstorage.QueryBuilder {
	return db.Build().Se("*").Fr(v.t())
}

//
//

func (v *Package) GetHookSecret() string {
	return v.hookSecret
}

func (v *Package) SetLatest(vers *Version) {
	vs := "v" + strconv.Itoa(vers.RealMajor) + "." + strconv.Itoa(vers.RealMinor)
	v.LatestVersion = vs
	doUp(v, "latest_version", "v"+strconv.Itoa(vers.RealMajor)+"."+strconv.Itoa(vers.RealMinor))
}

func (v *Package) GetVersion(major int, minor int) *Version {
	return Version{}.ByCode(v, major, minor)
}

func (v *Package) SetLicense(s string) {
	v.License = s
	doUp(v, "license", s)
}

func (v *Package) SetDescription(s string) {
	v.Description = s
	doUp(v, "description", s)
}

func (v *Package) SetStarCount(n int) {
	v.StarCount = n
	doUp(v, "star_count", strconv.Itoa(n))
}

File: src/db/Package.zig
const Package = @This();
pub const table_name = "packages";

const std = @import("std");
const string = []const u8;
const extras = @import("extras");

const _db = @import("./_db.zig");
const Version = _db.Version;
const Time = _db.Time;
const User = _db.User;
const Remote = _db.Remote;

const ox = @import("ox").sql;
const db = &ox.db;

id: u64 = 0,
uuid: ox.ULID,
owner: ox.ULID,
name: string,
created_on: Time,
remote: u64,
remote_id: string,
remote_name: string,
description: string,
license: string,
latest_version: string,
hook_secret: string,
star_count: u64,
clone_url: string,

pub fn create(alloc: std.mem.Allocator, owner: User, name: string, remote: Remote, rm_id: string, rm_name: string, desc: string, license: string, star_count: u64, clone_url: string) !Package {
    db.mutex.lock();
    defer db.mutex.unlock();

    return try ox.insert(alloc, &Package{
        .uuid = ox.factory.newULID(),
        .owner = owner.uuid,
        .name = name,
        .created_on = Time.now(),
        .remote = remote.id,
        .remote_id = rm_id,
        .remote_name = rm_name,
        .description = desc,
        .license = license,
        .latest_version = "",
        .hook_secret = try extras.randomSlice(alloc, std.crypto.random, u8, 16),
        .star_count = star_count,
        .clone_url = clone_url,
    });
}

pub usingnamespace ox.TableTypeMixin(Package);
pub usingnamespace ox.ByKeyGen(Package);
pub usingnamespace ox.JsonStructSkipMixin(@This(), &.{ "id", "hook_secret" });

pub const findVersionBy = ox.FindByGen(Package, Version, .p_for, .uuid).first;

pub fn latest(alloc: std.mem.Allocator) ![]const Package {
    return try db.collect(alloc, Package, "select * from packages order by id desc limit 15", .{});
}

pub fn topStarred(alloc: std.mem.Allocator) ![]const Package {
    return try db.collect(alloc, Package, "select * from packages order by star_count desc limit 15", .{});
}

pub fn versions(self: Package, alloc: std.mem.Allocator) ![]const Version {
    return try Version.byKeyAll(alloc, .p_for, self.uuid, .desc);
}

pub fn setLatest(self: *Package, alloc: std.mem.Allocator, vers: Version) !void {
    try self.update(alloc, .latest_version, try std.fmt.allocPrint(alloc, "{}", .{vers}));
}

pub fn getLatestValid(self: Package, alloc: std.mem.Allocator) !Version {
    return (try db.first(alloc, Version, "select * from versions where p_for = ? and (real_major > 0 or real_minor > 0) order by id desc", .{self.uuid})) orelse @panic("unreachable");
}

pub fn findVersionAt(self: Package, alloc: std.mem.Allocator, major: u32, minor: u32) !?Version {
    return try db.first(alloc, Version, "select * from versions where p_for = ? and real_major = ? and real_minor = ?", .{ self.uuid, major, minor });
}

pub fn cloneUrl(self: Package, alloc: std.mem.Allocator) !string {
    if (self.clone_url.len > 0) {
        return self.clone_url;
    }
    const repo = try _db.Remote.byKey(alloc, .id, self.remote);
    return try std.fmt.allocPrint(alloc, "https://{s}/{s}", .{ repo.?.domain, self.remote_name });
}
